README.md
# regta-period

**Library to make moment-independent periods in python.
It's designed especially for [Regta Framework](https://github.com/SKY-ALIN/regta), 
but with an ability to use it independently.**

[![versions](https://img.shields.io/pypi/pyversions/regta-period.svg)](https://github.com/SKY-ALIN/regta-period)
![Tests](https://github.com/SKY-ALIN/regta-period/actions/workflows/tests.yml/badge.svg)
![Code Quality](https://github.com/SKY-ALIN/regta-period/actions/workflows/code-quality.yml/badge.svg)
[![codecov](https://codecov.io/gh/SKY-ALIN/regta-period/branch/main/graph/badge.svg?token=NR7AKLXN5H)](https://codecov.io/gh/SKY-ALIN/regta-period)
[![PyPI version](https://badge.fury.io/py/regta-period.svg)](https://pypi.org/project/regta-period/)
[![license](https://img.shields.io/github/license/SKY-ALIN/regta-period.svg)](https://github.com/SKY-ALIN/regta-period/blob/main/LICENSE)

## Moment-Independence Idea Explanation

This term in this context means that relying on this approach we can get the time to time 
points regardless of the points in which we are.
```
        |-----------------|
        t1       t2     moment
--------|--------|--------|--------> time
                 |--------|
```

Whereas with the standard intervals like `datetime.timedelta`, we get an unnecessary offset:
```
        |-----------------|
        t1       t2     moment
--------|--------|--------|--------|--------> time
                 |-----------------|
```

For example, it is important in the context of the job scheduler, because when the
scheduler is redeployed or restarted, you can get an unnecessary time shift or
unnecessary execution of the job.

## Math Explanation Of Moment-Independence

### Regular Offset

Regular offset is the same as python's `timedelta` shift e.g. once per $n_1$ days,
$n_2$ hours, $n_3$ minutes, $n_4$ seconds, but with the moment-independence idea.

Essentially, it works as a remainder of the time division from the Unix epoch.
Let $t_{unix}$ be the moment of the Unix epoch, a moment that we can get a grip on.
$t$ is the current moment. Then, the time since epoch is:

$$\ \Delta t = t - t_{unix} $$

Let $T$ be our regular period. Thus, to calculate time until the next moment we must subtract
from our period the remainder of the division by the period. Final function to calculate
time until the next moment since current looks following:

$$\ f(t) = T - ( \Delta t \mod T ) = T - ( ( t - t_{unix} ) \mod T ) $$

### Time Offset

Time offset is stating the exact time e.g. at 9 pm, at 12 am, at 16:30, etc.
It works as a shift of the starting point in the exact time and time zone:

$$\ t_{unix} + \Delta t_{time} + \Delta t_{tz} $$

Note that it's not possible to combine exact time and short regular intervals such as
hours, minutes, and seconds.

### Time Windows

Time window is a static time frame in which the result should be included e.g. every Monday, every June, etc. 
A window may be from $t_{min}$ to $t_{max}$, then function result must be included in this interval:

$$ t + f(t) \in [t_{min}, t_{max}] $$

If the expression above is true, it means that the result is included in the time window, and the result is correct. 
If don't, we calculate the result from the maximum and calculate the next time window until we find a match:

$$ t + f(t) \notin [t_{min_n}, t_{max_n}] \longrightarrow f(t_{max_n}); [t_{min_{n+1}}, t_{max_{n+1}}] $$

## Installation

Install using `pip install regta-period` or `poetry add regta-period`

If you use python < 3.9, then also install backports: `pip install "backports.zoneinfo[tzdata]"`

## Examples

There are two ways to create periods: old school style and hipster style.

```python
from datetime import datetime
from zoneinfo import ZoneInfo
from regta_period import Period

# Hipster style
p = Period().every(3).days.at("17:00").by("Europe/Moscow")

# Old school style
p = Period(days=3, time="17:00", timezone=ZoneInfo("Europe/Moscow"))

# <Period: regular_offset=259200.0s, time_offset=61200s, timezone=Europe/Moscow>
# Every 3 days at 5 pm by Moscow time

t = datetime.now(tz=ZoneInfo("Europe/Moscow"))
next_moment: datetime = p.get_next(t)  # f(t) + t
```

You also may combine a few periods to a single object with the same interface:

```python
from datetime import datetime
from regta_period import Period, PeriodAggregation, Weekdays

# Hipster style
p = Period().on.weekdays.at("18:00") | Period().on.weekends.at("21:00")
# You also may replace `|` with `.OR` to write shorter and more human-readable code
p = Period().on.weekdays.at("18:00").OR.on.weekends.at("21:00")

# Old school style
p = PeriodAggregation(
    Period(
        weekdays=[Weekdays.MONDAY, Weekdays.TUESDAY, Weekdays.WEDNESDAY, Weekdays.THURSDAY, Weekdays.FRIDAY],
        time="18:00",
    ),
    Period(
        weekdays=[Weekdays.SATURDAY, Weekdays.SUNDAY],
        time="21:00",
    ),
)

# All of the above will the same result:
# <PeriodAggregation: <Period: regular_offset=86400.0s, time_offset=64800s, weekdays=Tuesday,Monday,Thursday,Wednesday,Friday> OR <Period: regular_offset=86400.0s, time_offset=75600s, weekdays=Sunday,Saturday>>
# At 6 pm on weekdays (Monday-Friday) and at 9 pm on weekends (Saturday-Sunday)

t = datetime.now()
timedelta_to_the_next_moment = p.get_interval(t)  # f(t)
```

---

Full documentation and reference are available at 
[regta-period.alinsky.tech](https://regta-period.alinsky.tech)


CHANGELOG.md
## 0.2.0 (28.12.2022)
* Add Python 3.11 support
* Add `AbstractPeriod.is_timezone_in_use`
* Add `Period.is_timezone_in_use`
* Add `PeriodAggregation.is_timezone_in_use`

## 0.1.0 (22.08.2022)
* Add `regta_period.AbstractPeriod`
  * `.get_next`
  * `.get_interval`
* Add `regta_period.Period`
  * `.every`
  * `.seconds`
  * `.minutes`
  * `.hours`
  * `.days`
  * `.hourly`
  * `.daily`
  * `.on`
  * `.monday`
  * `.tuesday`
  * `.wednesday`
  * `.thursday`
  * `.friday`
  * `.saturday`
  * `.sunday`
  * `.weekdays`
  * `.weekends`
  * `.at`
  * `.by`
  * `.get_next`
  * `.get_interval`
  * `.AND`
  * `.OR`
  * `.__add__`
  * `.__or__`
  * `.__repr__`
* Add `regta_period.PeriodAggregation`
  * `.get_next`
  * `.get_interval`
  * `.OR`
  * `.__or__`
  * `.__dir__`
  * `.__getattr__`
  * `.__repr__`
* Add `regta_period.Weekdays`
  * `.get`


tests/conftest.py
from datetime import datetime, timezone

try:
    import zoneinfo  # type: ignore
except ImportError:  # Backward compatibility for python < 3.9
    from backports import zoneinfo  # type: ignore

import pytest


@pytest.fixture
def unix():
    return datetime.utcfromtimestamp(0)


@pytest.fixture
def utc():
    return timezone.utc


@pytest.fixture
def utc7():
    return zoneinfo.ZoneInfo('Asia/Tomsk')


tests/test_enums.py
from datetime import datetime

from regta_period.enums import Weekdays


def test_weekdays():
    assert Weekdays.get(datetime(2022, 7, 24, 0, 0, 0)) == Weekdays.SUNDAY
    assert Weekdays.get(datetime(2022, 7, 25, 0, 0, 0)) == Weekdays.MONDAY
    assert Weekdays.get(datetime(2022, 7, 26, 0, 0, 0)) == Weekdays.TUESDAY
    assert Weekdays.get(datetime(2022, 7, 27, 0, 0, 0)) == Weekdays.WEDNESDAY
    assert Weekdays.get(datetime(2022, 7, 28, 0, 0, 0)) == Weekdays.THURSDAY
    assert Weekdays.get(datetime(2022, 7, 29, 0, 0, 0)) == Weekdays.FRIDAY
    assert Weekdays.get(datetime(2022, 7, 30, 0, 0, 0)) == Weekdays.SATURDAY


tests/test_at_time.py
from typing import Callable

from datetime import datetime

from regta_period import Period


def _assert_creation(f: Callable[..., Period], reverse: bool = False):
    try:
        f()
    except ValueError:
        assert reverse
    else:
        assert not reverse


def test_wrong_combination_of_time_and_interval():
    _assert_creation(lambda: Period().every(12).hours.at("12:00"), reverse=True)
    _assert_creation(lambda: Period().every(12).hours.at("24:00"), reverse=True)
    _assert_creation(lambda: Period().every(2).days.AND.every(12).hours.at("12:00"), reverse=True)
    _assert_creation(lambda: Period().every(2).days.AND.every(12).hours.at("00:00"), reverse=True)


def test_correct_combination_of_time_and_interval():
    _assert_creation(lambda: Period().every(2).days.at("12:00"))
    _assert_creation(lambda: Period().every(24).hours.at("12:12"))
    _assert_creation(lambda: Period().every(24).hours.at("24:00"))


def test_at_time(unix: datetime):
    p = Period().every(3).days.at("12:00")
    moment = datetime(year=1970, month=1, day=1, hour=12, minute=0, second=0)
    assert p.get_next(unix) == moment
    next_moment = datetime(year=1970, month=1, day=1+3, hour=12, minute=0, second=0)
    assert p.get_next(moment) == next_moment
    moment = next_moment
    next_moment = datetime(year=1970, month=1, day=1+6, hour=12, minute=0, second=0)
    assert p.get_next(moment) == next_moment

    p = Period().every(7).days.at("16:30:30")
    moment = datetime(year=1970, month=1, day=1, hour=16, minute=30, second=30)
    assert p.get_next(unix) == moment
    next_moment = datetime(year=1970, month=1, day=1+7, hour=16, minute=30, second=30)
    assert p.get_next(moment) == next_moment
    moment = next_moment
    next_moment = datetime(year=1970, month=1, day=1+14, hour=16, minute=30, second=30)
    assert p.get_next(moment) == next_moment


regta_period/periods.py
from typing import Dict, Iterable, Set, Tuple, Union

from abc import ABC, abstractmethod
from datetime import datetime, timedelta, timezone as datetime_timezone, tzinfo

try:
    import zoneinfo  # type: ignore
except ImportError:  # Backward compatibility for python < 3.9
    from backports import zoneinfo  # type: ignore

from .enums import Weekdays

utc = datetime_timezone.utc


class AbstractPeriod(ABC):
    """The minimum interface every period object has."""

    @abstractmethod
    def get_interval(self, dt: datetime) -> timedelta:
        """Get time to the next moment as timedelta since passed moment.

        Args:
            dt (datetime): Current moment (:math:`t`)

        Return:
            timedelta: Interval to the next moment (:math:`f(t)`)
        """
        raise NotImplementedError

    @abstractmethod
    def get_next(self, dt: datetime) -> datetime:
        """Get the next moment since passed moment.

        Args:
            dt (datetime): Current moment (:math:`t`)

        Return:
            datetime: The next moment (:math:`t + f(t)`)
        """
        raise NotImplementedError

    @property
    @abstractmethod
    def is_timezone_in_use(self) -> bool:
        """If timezone is specified, return True, else False."""
        raise NotImplementedError


class Period(AbstractPeriod):
    """The core logic of this module.

    Args:
        days (int): Amount of days for the regular offset.
        hours (int): Amount of hours for the regular offset.
        minutes (int): Amount of minutes for the regular offset.
        seconds (int): Amount of seconds for the regular offset.
        milliseconds (int): Amount of milliseconds for the regular offset.
        time (str): Exact time of moments (time offset). Format: "HH:MM" or "HH:MM:SS".
        timezone (Union[tzinfo, str, int, float]):
            Time zone for exact time.
            If :obj:`str`, then it will be converted into :obj:`tzinfo` via :class:`zoneinfo.ZoneInfo`.
            If :obj:`int` or :obj:`float`, then it will be used directly as an offset for the time offset.
        weekdays (Iterable[Weekdays]): Time windows of weekdays.
    """

    _every: int = 1
    _regular_offset: float = 0.0
    _time_offset: int = 0
    _timezone_offset: Union[int, None] = None
    _timezone: Union[tzinfo, None] = None
    _weekdays: Set[Weekdays]

    def __init__(
            self,
            days: int = 0,
            hours: int = 0,
            minutes: int = 0,
            seconds: int = 0,
            milliseconds: int = 0,
            time: Union[str, None] = None,
            timezone: Union[tzinfo, str, int, float, None] = None,
            weekdays: Union[Iterable[Weekdays], None] = None,
    ):
        self._regular_offset = (
            milliseconds * 0.001
            + seconds
            + minutes * 60
            + hours * 60 * 60
            + days * 60 * 60 * 24
        )
        if time is not None:
            self._set_time_offset(*self._parse_time(time))
        if timezone is not None:
            self._set_timezone(timezone)
        self._weekdays = set(weekdays) if weekdays is not None else set()

    def every(self, n: int) -> "Period":
        """Specify a factor for regular offset properties.

        Args:
            n (int): The factor.
        """
        self._every = n
        return self

    @property
    def seconds(self) -> "Period":
        """Seconds regular offset property. Must be used only after :attr:`.every` multiplier."""
        self._regular_offset += self._every
        return self

    @property
    def minutes(self) -> "Period":
        """Minutes regular offset property. Must be used only after :attr:`.every` multiplier."""
        self._regular_offset += self._every * 60
        return self

    @property
    def hours(self) -> "Period":
        """Hours regular offset property. Must be used only after :attr:`.every` multiplier."""
        self._regular_offset += self._every * 60 * 60
        return self

    @property
    def days(self) -> "Period":
        """Days regular offset property. Must be used only after :attr:`.every` multiplier."""
        self._regular_offset += self._every * 60 * 60 * 24
        return self

    @property
    def hourly(self) -> "Period":
        """Regular offset = every hour. The same as :code:`.every(1).hours`.
        Can't be combined with another regular offset.
        """
        if self._regular_offset:
            raise ValueError("Can't combine .hourly and other regular offset attributes")
        self._regular_offset = 60.0 * 60
        return self

    @property
    def daily(self) -> "Period":
        """Regular offset = every day. The same as :code:`.every(1).days`.
        Can't be combined with another regular offset.
        """
        if self._regular_offset:
            raise ValueError("Can't combine .daily and other regular offset attributes")
        self._regular_offset = 60.0 * 60 * 24
        return self

    @property
    def on(self) -> "Period":
        """This property does nothing. It's designed only to write better
        human-readable code, e.g. :code:`.on.monday`.
        """
        return self

    @property
    def monday(self) -> "Period":
        """Add Monday to the time windows list."""
        self._weekdays.add(Weekdays.MONDAY)
        return self

    @property
    def tuesday(self) -> "Period":
        """Add Tuesday to the time windows list."""
        self._weekdays.add(Weekdays.TUESDAY)
        return self

    @property
    def wednesday(self) -> "Period":
        """Add Wednesday to the time windows list."""
        self._weekdays.add(Weekdays.WEDNESDAY)
        return self

    @property
    def thursday(self) -> "Period":
        """Add Thursday to the time windows list."""
        self._weekdays.add(Weekdays.THURSDAY)
        return self

    @property
    def friday(self) -> "Period":
        """Add Friday to the time windows list."""
        self._weekdays.add(Weekdays.FRIDAY)
        return self

    @property
    def saturday(self) -> "Period":
        """Add Saturday to the time windows list."""
        self._weekdays.add(Weekdays.SATURDAY)
        return self

    @property
    def sunday(self) -> "Period":
        """Add Sunday to the time windows list."""
        self._weekdays.add(Weekdays.SUNDAY)
        return self

    @property
    def weekdays(self) -> "Period":
        """Add weekdays (Monday-Friday) to the time windows list."""
        self._weekdays.update({
            Weekdays.MONDAY,
            Weekdays.TUESDAY,
            Weekdays.WEDNESDAY,
            Weekdays.THURSDAY,
            Weekdays.FRIDAY,
        })
        return self

    @property
    def weekends(self) -> "Period":
        """Add weekends (Saturday-Sunday) to the time windows list."""
        self._weekdays.update({Weekdays.SATURDAY, Weekdays.SUNDAY})
        return self

    @staticmethod
    def _parse_time(time: str) -> Tuple[int, int, int]:
        values = tuple(map(int, time.split(":")))

        if len(values) == 2:
            hour, minute = values
            second = 0
        elif len(values) == 3:
            hour, minute, second = values
        else:
            raise ValueError(f"Wrong time format: {repr(time)}")

        return hour, minute, second

    def _set_time_offset(self, hour: int, minute: int, second: int) -> None:
        checking_delimiter = (24 * 60 * 60)
        if self._regular_offset % checking_delimiter:
            raise ValueError(
                "Can't combine .at method and too small regular offset. "
                "Don't combine attributes which are < day with .at time method."
            )

        self._time_offset = hour * 60 * 60 + minute * 60 + second

    def at(self, time: str) -> "Period":
        """Specify the moment exact time (time offset, :math:`\Delta t_{time}`).

        Args:
            time (str): Exact time. Format: "HH:MM" or "HH:MM:SS".
        """
        self._set_time_offset(*self._parse_time(time))
        return self

    def _set_timezone(self, timezone: Union[tzinfo, str, int, float]) -> None:
        if isinstance(timezone, tzinfo):
            self._timezone = timezone
        elif isinstance(timezone, str):
            self._timezone = zoneinfo.ZoneInfo(timezone)
        elif isinstance(timezone, (int, float)):
            self._timezone_offset = int(timezone * 60 * 60)
        else:
            raise ValueError("Unsupported timezone type")

    def by(self, timezone: Union[tzinfo, str, int, float]) -> "Period":
        """Specify the time zone for the exact time.

        Args:
            timezone (Union[tzinfo, str, int, float]):
                Time zone.
                If :obj:`str`, then it will be converted into :obj:`tzinfo` via :class:`zoneinfo.ZoneInfo`.
                If :obj:`int` or :obj:`float`, then it will be used directly as an offset for the time offset.
        """
        self._set_timezone(timezone)
        return self

    def _get_initial_datetime(self) -> datetime:
        if self._timezone is not None:
            return datetime.fromtimestamp(self._time_offset, tz=utc).replace(tzinfo=self._timezone)
        if self._timezone_offset is not None:
            return datetime.fromtimestamp(self._time_offset - self._timezone_offset, tz=utc)
        return datetime.utcfromtimestamp(self._time_offset)

    def get_next(self, dt: datetime) -> datetime:
        delta_t = (dt - self._get_initial_datetime()).total_seconds()
        # if _regular_offset is not specified, calculate as .daily
        regular_offset = self._regular_offset or 60.0 * 60 * 24
        next_seconds = regular_offset - (delta_t % regular_offset)

        res = dt + timedelta(seconds=next_seconds)

        if self._weekdays and Weekdays.get(res) not in self._weekdays:
            return self.get_next(res)

        return res

    def get_interval(self, dt: datetime) -> timedelta:
        return self.get_next(dt) - dt

    @property
    def is_timezone_in_use(self) -> bool:
        return self._timezone is not None or self._timezone_offset is not None

    @property
    def AND(self) -> "Period":
        """This property does nothing. It's designed only to write better
        human-readable code, e.g. :code:`.on.monday.AND.tuesday`.
        It's uppercase because :code:`and` is a reserved word in python.
        """
        return self

    @property
    def OR(self) -> "PeriodAggregation":
        """Create :class:`PeriodAggregation` from this period and a new empty period.

        It's designed to write better human-readable code,
        e.g. :code:`.on.weekdays.at("18:00").OR.on.weekends.at("21:00")`.
        It's uppercase because :code:`or` is a reserved word in python.
        """
        return PeriodAggregation(self, Period())

    def __add__(self, other: "Period") -> "Period":
        """Combine periods as a sum of regular offset and time windows.
        Can't sum objects with a different time offset and time zone.
        """
        if self._time_offset != other._time_offset:
            raise ValueError(
                "Can't sum periods with a different time. "
                "Hint: try to use | instead"
            )
        if self._timezone != other._timezone or self._timezone_offset != other._timezone_offset:
            raise ValueError(
                "Can't sum periods with a different timezone. "
                "Hint: try to use an operator `|` or property `.OR` instead"
            )
        self._regular_offset += other._regular_offset
        self._weekdays.update(other._weekdays)
        return self

    def __or__(self, other: Union["Period", "PeriodAggregation"]) -> "PeriodAggregation":
        """Create new :class:`PeriodAggregation` from this period and the passed period."""
        if isinstance(other, Period):
            return PeriodAggregation(self, other)
        return PeriodAggregation(self, *other.periods)

    def __repr__(self):
        data: Dict[str, str] = {
            "regular_offset": f"{self._regular_offset or 60.0 * 60 * 24}s",
            "time_offset": f"{self._time_offset}s",
        }
        if self._timezone is not None:
            data["timezone"] = str(self._timezone)
        elif self._timezone_offset is not None:
            data["timezone_offset"] = f"{self._timezone_offset}s"
        if self._weekdays:
            data["weekdays"] = ",".join(map(lambda x: x.name.capitalize(), self._weekdays))

        data_str = ", ".join(f"{key}={value}" for key, value in data.items())
        return f"<{self.__class__.__name__}: {data_str}>"


class PeriodAggregation(AbstractPeriod):
    """Aggregation class for :class:`Period`.

    It contains the logic of how to get the nearest time moment and add data into
    the last period object.

    Args:
        *periods (Tuple[Period]): Periods to aggregate.

    Attributes:
        periods (Tuple[Period]): Aggregated periods.
    """

    def __init__(self, *periods: Period):
        if not periods:
            raise ValueError("No period has been passed")
        self.periods = periods

    def get_next(self, dt: datetime) -> datetime:
        return min(map(lambda period: period.get_next(dt), self.periods))

    def get_interval(self, dt: datetime) -> timedelta:
        return self.get_next(dt) - dt

    @property
    def is_timezone_in_use(self) -> bool:
        return any(map(lambda period: period.is_timezone_in_use, self.periods))

    @property
    def OR(self) -> "PeriodAggregation":
        """Create a new :class:`PeriodAggregation` from these periods of
        aggregation and a new empty period.

        It's designed to write better human-readable code,
        e.g. :code:`.on.weekdays.at("18:00").OR.on.weekends.at("21:00")`.
        It's uppercase because :code:`or` is a reserved word in python.
        """
        return PeriodAggregation(*self.periods, Period())

    def __or__(self, other: Union["Period", "PeriodAggregation"]) -> "PeriodAggregation":
        """Create a new :class:`PeriodAggregation` from these periods of
        aggregation and the passed period."""
        if isinstance(other, Period):
            return PeriodAggregation(*self.periods, other)
        return PeriodAggregation(*self.periods, *other.periods)

    def __dir__(self) -> Iterable[str]:
        """Extended implementation of the standard.
        It adds attributes of :class:`Period` class.
        """
        return sorted(set(dir(PeriodAggregation)) | set(dir(Period)))

    def __getattr__(self, attr):
        """Return an attribute of the last period in the list of periods with
        a wrap to return this object instead of the last period.
        """
        _attr = getattr(self.periods[-1], attr)

        if isinstance(getattr(Period, attr), property):
            return self

        def wrapper(*args, **kwargs):
            _attr(*args, **kwargs)
            return self

        return wrapper

    def __repr__(self):
        return f"<{self.__class__.__name__}: {' OR '.join(map(repr, self.periods))}>"
